pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/RoutingManager.java

```diff
@@ -131,44 +131,42 @@ private void processExternalViewChange() {
     Stat[] stats = _zkDataAccessor.getStats(externalViewPaths, AccessOption.PERSISTENT);
     long fetchStatsEndTimeMs = System.currentTimeMillis();
 
-    int numRoutingEntriesToUpdate = 0;
+    List<String> tablesToUpdate = new ArrayList<>();
     for (int i = 0; i < numTables; i++) {
       Stat stat = stats[i];
       if (stat != null) {
         RoutingEntry routingEntry = routingEntries.get(i);
         if (stat.getVersion() != routingEntry.getLastUpdateExternalViewVersion()) {
-          numRoutingEntriesToUpdate++;
+          String tableNameWithType = routingEntry.getTableNameWithType();
+          tablesToUpdate.add(tableNameWithType);
           try {
-            updateRoutingEntryOnExternalViewChange(routingEntry);
+            ExternalView externalView = getExternalView(tableNameWithType);
+            if (externalView == null) {
+              LOGGER.warn("Failed to find external view for table: {}, skipping updating routing entry",
+                  tableNameWithType);
+              continue;
+            }
+            Set<String> onlineSegments = getOnlineSegments(tableNameWithType);
+            if (onlineSegments == null) {
+              LOGGER
+                  .warn("Failed to find ideal state for table: {}, skipping updating routing entry", tableNameWithType);
+              continue;
+            }
+            routingEntry.onExternalViewChange(externalView, onlineSegments);
           } catch (Exception e) {
             LOGGER
                 .error("Caught unexpected exception while updating routing entry on external view change for table: {}",
-                    routingEntry.getTableNameWithType(), e);
+                    tableNameWithType, e);
           }
         }
       }
     }
     long updateRoutingEntriesEndTimeMs = System.currentTimeMillis();
 
     LOGGER.info(
-        "Processed external view change in {}ms (fetch {} external view stats: {}ms, update {} routing entries: {}ms)",
+        "Processed external view change in {}ms (fetch {} external view stats: {}ms, update {} routing entries ({}): {}ms)",
```

Comment:
It is not intuitive from this message as to how many tables the routing table was updated for? Is tablesToUpdate.size() this number?

May be we should make update N routing entries to update N routing table entries for M tables.

Reply:
Changed it to `..., update routing entry for 2 tables ([foo_OFFLINE, bar_REALTIME]): 5ms`